### ISSUE TITLE
Disable libcdb code

### ISSUE THREAD

**Issue by zachriggle** [Gallopsled/pwntools#1171]
My libc database hasn't been maintained in a year, and the code should be either removed or set to default-off.

**Comment by david942j** [comment]
May I know why stop maintaining libcdb?

**Comment by zachriggle** [comment]
It requires a LOT of storage and there are so many different cloud
distributions nowadays that all have different libc binaries... it's not
really worth it.

You can see the current state here: https://gitlab.com/libcdb/libcdb

Note that it requires about 30GB to check out.  It looks like it hasn't
been updated in a ~year.

*Zach Riggle*

**Comment by Arusekk** [comment]
It should be possible to store libc's from a couple most popular distros (ubuntu, debian, arch, fedora) from e.g. last 6 months. Should not take too much space.

Also, libcdb could try to check local libc (or libc on SSH connection) if its build ID happens to match the one being exploited (could be disabled). It is quite a common situation and sometimes exploits are executed locally, the cost of checking local libc's build ID is near-zero.

**Comment by zachriggle** [comment]
@Arusekk I already have a repository with scripts to automatically scrape all libc from most popular distros: https://gitlab.com/libcdb/libcdb

This is the repo that the DynELF module makes use of, and that I was referring to above.  

It's organized by build-ID, sha1, sha256, and md5:
https://github.com/Gallopsled/pwntools/blob/94647d6c10fdf8a395c61641eaff7ab5aecfcede/pwnlib/libcdb.py#L53

It includes both the dynamically-linked library (libc.so) as well as the statically linked library (libc.a) since the latter is useful for FLIRT signatures.

All that needs to happen is for somebody (not me, I cannot contribute) to clone the repo and run the update scripts (possibly need some fixing up too).  I can grant push access to the repo.  You'll need about 50GB of free space.

**Comment by zachriggle** [comment]
Also, when connecting via SSH, you can just download all of the libraries without worrying about build ID.

```py
>>> s=ssh('level1', 'io.netgarage.org', password='level1')
[x] Connecting to io.netgarage.org on port 22
[+] Connecting to io.netgarage.org on port 22: Done
>>> s.libs('/bin/cat')
```

```
$ tree io.netgarage.org
io.netgarage.org
├── bin
│   └── cat
└── lib
    ├── i386-linux-gnu
    │   └── libc.so.6
    └── ld-linux.so.2
```

Or alternately:

```py
>>> io = s.process('sh')
>>> io.libs()
```

**Comment by ajabep** [comment]
Hello,

Thank you for having maintained your database during all this time :)

What I see is that:

* We have a module which is useful for full-remote CTF challenge.
* This tool uses an out-to-date database, thus becomes less and less relevant.
* Some people have built some databases online (https://libc.rip, https://libc.blukat.me, etc.) which are up-to-date. This happens whenever a database becomes too outdated.

May I make the following suggestions (I can submit PR to do so):

1. Add a way to change the service which is used.  
   The service can have a similar API or can implement a specific API to do so.  
   Change the website which is used is useful: it allows the community to build services with their specificity (some repo specialized in an arch, some with only static libs, some with only some OS, etc.). The proliferation and specialization of databases can reduce to come again in the current situation and making people imagining solutions and sharing them.  
   Databases with specificity are already a thing that we can see for some databases (e.g. for FLIRT, we can find on GitHub https://github.com/push0ebp/sig-database or https://github.com/Maktm/FLIRTDB).

2. Allowing the download of compressed files. This reduce the download data and the disk size of the DB.

3. "Standardize" an API, or implement one if some exists, to interact with the database. The one which is currently implemented is great, allowing the storage via a git repo.

   [![](https://imgs.xkcd.com/comics/standards.png)](https://xkcd.com/927/)

What are your thoughts about it?

**Comment by zachriggle** [comment]
I wouldn't mind adding support for other databases, as long as they provide lookup by Build-ID.  

The code that's `libcdb`-specific is really small, and the scenario where libcdb is most useful is when used with DynELF, which uses the Build-ID lookup.

**Comment by Arusekk** [comment]
I guess it is closed by #1704.
